src/crimson/os/seastore/transaction_manager.h

```diff
@@ -362,6 +362,71 @@ class TransactionManager : public AsyncCleaner::ExtentCallbackInterface {
     });
   }
 
+  /**
+   * map_existing_extent
+   *
+   * Allocates a new extent at given existing_paddr that must be absolute and
+   * reads disk to fill the extent.
+   * The common usage is that remove the LogicalCachedExtent (laddr~length at paddr)
+   * and map extent to multiple new extents.
+   * placement_hint and generation should follow the original extent.
+   */
+  using map_existing_extent_iertr =
+    alloc_extent_iertr::extend_ertr<Device::read_ertr>;
+  template <typename T>
+  using map_existing_extent_ret =
+    map_existing_extent_iertr::future<TCachedExtentRef<T>>;
+  template <typename T>
+  map_existing_extent_ret<T> map_existing_extent(
```

Comment:
Seems to me this logic circumvents the transaction conflict mechanisms because the remapped extent is read out independently without the original extent.

Things will go wrong if conflicts should happen but the mechanism is bypassed. For example, when the reclaiming transaction moves away the orginal extent at the same time.

Am I missing anything?

Reply:
remapping an extent needs to retire it first, which can sync with other transactions.

https://github.com/ceph/ceph/blob/f9f1e60c0929ea2332941a585f8c27aa608ae92a/src/test/crimson/seastore/test_transaction_manager.cc#L759-L762
https://github.com/ceph/ceph/blob/f88e1d2933dcb00b0e2dbe82cf62fbcbf05a8422/src/crimson/os/seastore/transaction_manager.h#L370-L371

Reply:
I can add an assertion to avoid bad usage:
```cpp
  bool Transaction::is_retired(laddr_t addr, extent_len_t len) {
    // look up retired_set
  }

  map_existing_extent_ret<T> map_existing_extent(
    Transaction &t,
    laddr_t laddr_hint,
    paddr_t existing_paddr,
    extent_len_t length,
    placement_hint_t placement_hint = placement_hint_t::HOT,
    reclaim_gen_t gen = DIRTY_GENERATION) {
  assert(t.is_retired(laddr_hint, length));
  //...
  }
```